src/componentSetBuilder.ts

```diff
@@ -38,64 +37,69 @@ export class ComponentSetBuilder {
     try {
       if (sourcepath) {
         logger.debug(`Building ComponentSet from sourcepath: ${sourcepath.toString()}`);
+        const fsPaths: string[] = [];
         sourcepath.forEach((filepath) => {
-          if (fs.fileExistsSync(filepath)) {
-            csAggregator.push(...ComponentSet.fromSource(path.resolve(filepath)));
-          } else {
+          if (!fs.fileExistsSync(filepath)) {
             throw new SfdxError(`The sourcepath "${filepath}" is not a valid source file path.`);
           }
+          fsPaths.push(path.resolve(filepath));
         });
+        componentSet = ComponentSet.fromSource({ fsPaths });
       }
 
       // Return empty ComponentSet and use packageNames in the library via `.retrieve` options
       if (packagenames) {
         logger.debug(`Building ComponentSet for packagenames: ${packagenames.toString()}`);
-        csAggregator.push(...new ComponentSet([]));
+        componentSet ??= new ComponentSet();
       }
 
       // Resolve manifest with source in package directories.
       if (manifest) {
         logger.debug(`Building ComponentSet from manifest: ${manifest.manifestPath}`);
         const directoryPaths = options.manifest.directoryPaths;
         logger.debug(`Searching in packageDir: ${directoryPaths.join(', ')} for matching metadata`);
-        const compSet = await ComponentSet.fromManifest({
+        componentSet = await ComponentSet.fromManifest({
           manifestPath: manifest.manifestPath,
           resolveSourcePaths: options.manifest.directoryPaths,
           forceAddWildcards: true,
         });
-        csAggregator.push(...compSet);
       }
 
       // Resolve metadata entries with source in package directories.
       if (metadata) {
         logger.debug(`Building ComponentSet from metadata: ${metadata.metadataEntries.toString()}`);
         const registry = new RegistryAccess();
+        const compSetFilter = new ComponentSet();
+        componentSet ??= new ComponentSet();
 
         // Build a Set of metadata entries
-        const filter = new ComponentSet();
-        metadata.metadataEntries.forEach((entry) => {
-          const splitEntry = entry.split(':');
-          // try and get the type by name to ensure no typos or errors in type name
-          // matches toolbelt functionality
+        metadata.metadataEntries.forEach((rawEntry) => {
+          const splitEntry = rawEntry.split(':');
+          // The registry will throw if it doesn't know what this type is.
           registry.getTypeByName(splitEntry[0]);
-          filter.add({
+          const entry = {
             type: splitEntry[0],
             fullName: splitEntry.length === 1 ? '*' : splitEntry[1],
-          });
+          };
+          // Add to the filtered ComponentSet for resolved source paths,
+          // and the unfiltered ComponentSet to build the correct manifest.
+          compSetFilter.add(entry);
```

Comment:
and this is only required for the `metadata` option

Reply:
Because for the `sourcepath` option you are only working with components in the project not the org.  The `manifest` option handles this for us with `ComponentSet.fromManifest()`.  This code is nearly identical to the code in `fromManifest()` and that's where this really belongs (in SDR) but that's a separate work item.